src/env/hedged/mod.rs

```diff
@@ -0,0 +1,671 @@
+/// because rewrite is a background job that doesn’t affect foreground latency.
+///
+/// For read, for performance consideration, it doesn't serialize read
+/// operations to disks' channel. It just reads from the disk which has handled
+/// task of larger sequence number.
+///
+/// But what if it's blocking for a long time or down, then the infinite
+/// accumulated operations record would exhaust the memory causing OOM. To avoid
+/// that, once the channel piles up to some extent, it just abandons that disk,
+/// which sends an `Pause` task to the channel and does not send tasks to that
+/// disk's channel anymore, while operations on the rewrite files' is not paused
+/// and still wait on both disks. So it has the chance that one disk written a
+/// rewrite entry while the origin entry is not written yet. But it's fine as
+/// rewrite read is performed on the faster disk which has the origin entry.
+///
+/// And later on, when the blocked disk comes back to normal, it would
+/// finally consume to the end of channel where is the `Pause` task indicating
+/// the disk is recovered. Then the slow disk would request a snapshot from
+/// faster disk, let the normal disk copy all the append files to that disk.
+/// And new changes resume being sent to the disk's channel. With that, all the
+/// writes to the new mutable append file during the copy won't be lost, and
+/// two disks can be synced eventually.
+///
+/// As the raft log file number increases monotonically, after a restart, it can
+/// check the latest append(exclude rewrite files) file number to decide
+/// which disk has the latest data. If the largest file numbers are the same, we
+/// can't simply use the file size to decide due to file recycling. Instead,
+/// scan the two files to find the biggest offset of the latest and valid record
+/// in them. And finally, regarding the disk of the bigger one as the latest.
+/// The latest one is responsible to sync the missing append files to the other
+/// disk.
+///
+/// It relays on some raft-engine invariants:
+/// 1. Raft log is append only.
+/// 2. Raft log is read-only once it's sealed.
+/// Here is the TLA+ proof https://github.com/pingcap/tla-plus/pull/41
+
+// TODO: handle specially on config change(upgrade and downgrade)
+// TODO: fallback to one disk, if the other disk is down for a long time and
+// close to full
+// TODO: consider encryption
+
+// All operations of file system trait are sent to the channels of both disks
+// through `HedgedFileSystem`.
+pub struct HedgedFileSystem {
+    base: Arc<DefaultFileSystem>,
+
+    path1: PathBuf,
+    path2: PathBuf,
+
+    sender: HedgedSender,
+
+    seqno1: Arc<AtomicU64>,
+    seqno2: Arc<AtomicU64>,
+
+    thread1: Option<JoinHandle<()>>,
+    thread2: Option<JoinHandle<()>>,
+}
+
+impl HedgedFileSystem {
+    pub fn new(base: Arc<DefaultFileSystem>, path1: PathBuf, path2: PathBuf) -> Self {
+        let (tx1, rx1) = unbounded::<(SeqTask, Callback)>();
```

Comment:
Should we limit the length of the channel to prevent the OOM, which is the hard limit  HedgedFileSystem can use